File: ISOnDemandTableView/ISOnDemandTableView.m
//
//  ISOnDemandTableView.m
//  ISOnDemandTableView
//
//  Created by Dielson Sales on 21/12/16.
//  Copyright © 2016 Ilhasoft. All rights reserved.
//

#import "ISOnDemandTableView.h"
#import "ISOnDemandTableViewCell.h"

#define SYSTEM_VERSION_GRATERTHAN_OR_EQUALTO(v)  ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] != NSOrderedAscending)

@implementation ISOnDemandTableView {
    UIRefreshControl *compatRefreshControl;
}

@synthesize interactor;

# pragma mark - Constructors

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self initialize];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initialize];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self initialize];
    }
    return self;
}

- (void)initialize
{
    compatRefreshControl = [[UIRefreshControl alloc] init];
    if (SYSTEM_VERSION_GRATERTHAN_OR_EQUALTO(@"10.0")) {
        self.refreshControl = compatRefreshControl;
    } else {
        // Fallback on earlier versions
        [self insertSubview:compatRefreshControl atIndex:0];
    }
    [compatRefreshControl addTarget:self action:@selector(onPullToRefresh) forControlEvents:UIControlEventValueChanged];
    self.dataSource = self;
    self.delegate = self;
}

- (void)setInteractor:(ISOnDemandTableViewInteractor *)newInteractor
{
    self->interactor = newInteractor;
    self->interactor.delegate = self;
}

- (void)onPullToRefresh
{
    //[self setFooterSpinner];
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandWillStartLoading:)]) {
        [self.onDemandTableViewDelegate onDemandWillStartLoading:self];
    }
    
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandWasPulledToRefresh:)]) {
        [self.onDemandTableViewDelegate onDemandWasPulledToRefresh:self];
    }
    
    [self.interactor refreshAllContent];
}

- (void)setFooterSpinner
{
    UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0.0, 0.0, [UIScreen mainScreen].bounds.size.width, 30.0)];
    [spinner startAnimating];
    spinner.color = [UIColor grayColor];
    self.tableFooterView = spinner;
}

- (void)loadContent
{
    if (self.onDemandTableViewDelegate == nil) {
        [NSException raise:@"ISOnDemandTableViewDelegateNotSet" format:@"You must set the ISOnDemandTableViewDelegate before calling loadContent"];
    }
    
    if (!self.interactor.isFetching && !_ignoreLoadRequests) {
        if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandWillStartLoading:)]) {
            [self.onDemandTableViewDelegate onDemandWillStartLoading:self];
        }
        [self.interactor loadItems];
        
        if (self.interactor.hasMoreItems && self.showFooterSpinner) {
            self.showFooterSpinner = NO;
            [self setFooterSpinner];
        }
    }
}

# pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.interactor.objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *reuseIdentifier = [self.onDemandTableViewDelegate onDemandTableView:self reuseIdentifierForCellAtIndexPath:indexPath];
    UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:reuseIdentifier forIndexPath:indexPath];
    if ([cell conformsToProtocol:@protocol(ISOnDemandTableViewCell)]) {
        UITableViewCell<ISOnDemandTableViewCell> *onDemandTableViewCell = (UITableViewCell<ISOnDemandTableViewCell> *)cell;
        [onDemandTableViewCell setupCellWithObject:[self.interactor.objects objectAtIndex:indexPath.row] atIndexPath:indexPath];
    }
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:setupCell:atIndexPath:)]) {
        [self.onDemandTableViewDelegate onDemandTableView:self setupCell:cell atIndexPath:indexPath];
    }
    return (UITableViewCell *)cell;
}

# pragma mark - UIScrollView

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:scrollViewDidScroll:)]) {
        [self.onDemandTableViewDelegate onDemandTableView:self scrollViewDidScroll:scrollView];
    }
}

# pragma mark - UITableViewDelegate

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    CGFloat contentYOffset = self.contentOffset.y + self.frame.size.height;
    if (contentYOffset >= self.contentSize.height) {
        NSLog(@"Reached end of tableview");
        self.showFooterSpinner = YES;
        [self loadContent];
    }
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
    if (decelerate == NO) {
        CGFloat contentYOffset = self.contentOffset.y + self.frame.size.height;
        if (contentYOffset >= self.contentSize.height) {
            NSLog(@"Reached end of tableview");
            self.showFooterSpinner = YES;
            [self loadContent];
        }
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:didSelectRowAtIndexPath:)]) {
        [self.onDemandTableViewDelegate onDemandTableView:self didSelectRowAtIndexPath:indexPath];
    }
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:cell:willAppearAtIndexPath:)]) {
        [self.onDemandTableViewDelegate onDemandTableView:self cell:cell willAppearAtIndexPath:indexPath];
    }
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:cell:willDisappearAtIndexPath:)]) {
        [self.onDemandTableViewDelegate onDemandTableView:self cell:cell willDisappearAtIndexPath:indexPath];
    }
}

-(UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:viewForHeaderInSection:)]) {
        return [self.onDemandTableViewDelegate onDemandTableView:self viewForHeaderInSection:&section];
    } else {
        return NULL;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:heightForHeaderAtSection:)]) {
        return [self.onDemandTableViewDelegate onDemandTableView:self heightForHeaderAtSection:&section];
    } else {
        return UITableViewAutomaticDimension;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if ([self.onDemandTableViewDelegate respondsToSelector:@selector(onDemandTableView:heightForRowAtIndexPath:)]) {
        return [self.onDemandTableViewDelegate onDemandTableView:self heightForRowAtIndexPath:indexPath];
    } else {
        return UITableViewAutomaticDimension;
    }
}

# pragma mark - ISOnDemandTableViewInteractorDelegate

- (void)onObjectsFetched:(NSArray *)lastObjects error:(NSError *)error
{
    self.tableFooterView = nil;
    if (compatRefreshControl.isRefreshing) {
        [compatRefreshControl endRefreshing];
    }
    [self.onDemandTableViewDelegate onDemandTableView:self onContentLoad:lastObjects withError:error];
    if (error == nil) {
        [self reloadData];
    }

}

@end

File: ISOnDemandTableView/ISOnDemandTableView.swift
//
//  ISOnDemandTableView.swift
//  ISOnDemandTableView
//
//  Created by Yves Bastos on 05/11/2017.
//  Copyright © 2017 Ilhasoft. All rights reserved.
//

import UIKit

class ISOnDemandTableView: UITableView {
    var onDemandDelegate: ISOnDemandTableViewDelegate?
    var interactor: ISOnDemandTableViewInteractor! {
        didSet {
            interactor?.delegate = self
        }
    }
    var isLoadingContent = false
    
    //MARK: Init
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        initialize()
    }
    
    override init(frame: CGRect, style: UITableViewStyle) {
        super.init(frame: frame, style: style)
        initialize()
    }
    
    fileprivate func initialize() {
        if #available(iOS 10.0, *) {
            refreshControl = UIRefreshControl()
            refreshControl?.addTarget(self, action: #selector(onPullToRefresh), for: .valueChanged)
        } 
        self.dataSource = self
        self.delegate = self
    }
    
    //MARK: Class Methods
    @objc fileprivate func onPullToRefresh() {
        onDemandDelegate?.onDemandWasPulled(toRefresh: self)
        interactor?.refreshAllContent()
    }
    
    /**
     Loads the contents in the on demand tableView.
     */
    func loadContent() {
        guard let _ = onDemandDelegate, let interactor = interactor else {
            fatalError("You must set both ISOnDemandColectionViewDelegate and ISOnDemandCollectionViewInteractor before calling loadContent")
        }
        
        if !interactor.isFetching {
            interactor.loadItems()
            setFooterSpinner(to: true)//interactor.hasMoreItems)
        }
    }
    
    //MARK: Util
    func setFooterSpinner(to show: Bool) {
        //TODO:
    }
}

extension ISOnDemandTableView: ISOnDemandTableViewInteractorDelegate {
    func onObjectsFetched(_ objects: [Any]?, _ error: Error?) {
        self.setFooterSpinner(to: false)
        if #available(iOS 10.0, *) {
            self.refreshControl?.endRefreshing()
        }
        
        self.onDemandDelegate?.onDemandTableView(self, onContentLoad: objects, withError: error)
        
        if error == nil {
            self.reloadData()
        }
    }
}

extension ISOnDemandTableView: UITableViewDataSource, UITableViewDelegate {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return interactor.objects.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = onDemandDelegate?.onDemandTableView(self, reuseIdentifierForCellAt: indexPath) ?? ""
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)
        (cell as? ISOnDemandTableViewCell)?.setup(with: interactor.objects[indexPath.row], at: indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return onDemandDelegate?.onDemandTableView(self, heightForRowAt: indexPath) ?? UITableViewAutomaticDimension
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        onDemandDelegate?.onDemandTableView(self, didSelectRowAt: indexPath)
    }
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        onDemandDelegate?.onDemandTableView(self, cell: cell, willAppearAt: indexPath)
    }
    
    func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        onDemandDelegate?.onDemandTableView(self, cell: cell, willDisappearAt: indexPath)
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return onDemandDelegate?.onDemandTableView(self, heightForHeaderAtSection: section) ?? UITableViewAutomaticDimension
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return onDemandDelegate?.onDemandTableView(self, viewForHeaderAtSection: section)
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        onDemandDelegate?.onDemandTableViewDidScroll(self)
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        let contentYOffset = self.contentOffset.y + self.frame.size.height
        if contentYOffset >= self.contentSize.height {
            NSLog("Reached end of tableview")
            setFooterSpinner(to: true)
            loadContent()
        }
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        guard !decelerate else {return}
      
        let contentYOffset = self.contentOffset.y + self.frame.size.height
        if contentYOffset >= self.contentSize.height {
            NSLog("Reached end of tableview")
            setFooterSpinner(to: true)
            loadContent()
        }
    }
}

